src/engine/enginemaster.cpp

```diff
@@ -443,48 +530,212 @@ void EngineMaster::process(const int iBufferSize) {
     }
 
     if (masterEnabled) {
-        // Mix the three channels together. We already mixed the busses together
-        // with the channel gains and overall master gain.
-        if (!m_pMasterTalkoverMix->toBool()) {
-            // Add Talkover to Master output
-            SampleUtil::copy4WithGain(m_pMaster,
-                    m_pOutputBusBuffers[EngineChannel::LEFT], 1.0,
-                    m_pOutputBusBuffers[EngineChannel::CENTER], 1.0,
-                    m_pOutputBusBuffers[EngineChannel::RIGHT], 1.0,
+        // Mix the crossfader orientation buffers together into the master mix
+        SampleUtil::copy3WithGain(m_pMaster,
+            m_pOutputBusBuffers[EngineChannel::LEFT], 1.0,
+            m_pOutputBusBuffers[EngineChannel::CENTER], 1.0,
+            m_pOutputBusBuffers[EngineChannel::RIGHT], 1.0,
+            iBufferSize);
+
+        // Process master, booth, and record/broadcast buffers according to the
+        // TalkoverMixMode configured in DlgPrefSound
+
+        // Mixxx gets input signals delayed by the input latency,
+        // so to preserve the relative timing of the input with Mixxx's outputs
+        // (headphones, master, booth, and record/broadcast), delay the outputs
+        // by the input latency before mixing the inputs into them.
+
+        // This way, if the microphone user is on beat with the output they hear from
+        // the speakers/headphones, the input will remain on beat at the expense of
+        // adding some latency to Mixxx's output (no latency will be added to the
+        // microphone signal though). If no microphone inputs are configured, avoid adding
+        // this extra latency.
+
+        // If the input signal is being directly monitored (mixed with the master output
+        // in hardware), do not delay the audible outputs. However, delay the
+        // record/broadcast signal to keep it on beat and consistent with what
+        // the user hears from the speakers/headphones.
+        if (configuredTalkoverMixMode == TalkoverMixMode::MASTER) {
+            // Apply input latency compensation to master mix
+            // Do not add unnecessary latency if no microphones are configured
+            if (m_pNumMicsConfigured->get() > 0) {
+                m_pInputLatencyCompensationDelay->process(m_pMaster, iBufferSize);
+            }
+
+            // Copy master mix to booth output with booth gain before mixing
+            // talkover with master mix
+            if (boothEnabled) {
+                CSAMPLE boothGain = m_pBoothGain->get();
+                if (m_bRampingGain) {
+                    SampleUtil::copy1WithRampingGain(m_pBooth, m_pMaster,
```

Comment:
> It really looks like, we have different ideas from latency compensation. I was not able to test is up to now, because the comensation delay was always added on to the headphones and booth. What is your understanding of latency compensation? What is the problem you suffer without it and how does you implementation solves the issue?

I described the issue this is solving [above](https://github.com/mixxxdj/mixxx/pull/1279#issuecomment-307509949):
> Previously, without any latency compensation, if the performer was on beat with what they were hearing, the master/booth output would be off beat, which is very disorienting and makes it pretty much impractical to perform.

It has always been possible to work around this with direct monitoring, but then the performance could not be recorded or broadcasted as heard. That also excludes the possibility of using Mixxx's effects on the live input.

Delaying the signals generated within Mixxx by half the round trip latency before mixing in the live inputs, which are inevitably delayed, aligns the mix of the mics and the software generated signals. Now, when I count or play into a microphone on beat with what I hear from the master output, when that input gets mixed into the master output, it will be on beat when I hear it come back out the master output. Inevitably I hear it after the round trip latency, but the important aspect is that when I am on beat, I hear myself on beat. When I play off beat, I hear how off beat I am.

This has nothing to do with speaker placement. You do not need a headphone or booth output to experience this issue. All you need is a microphone or other live input and the master output.